include/seqan3/search/fm_index/fm_index_cursor.hpp

```diff
@@ -482,9 +482,25 @@ class fm_index_cursor
                       "The alphabet types of the given text and index differ.");
         assert(index != nullptr);
 
-        size_type const loc = offset() - index->index[node.lb];
-        size_type const query_begin = loc - index->text_begin_rs.rank(loc + 1) + 1; // Substract delimiters
-        return text | views::join | views::slice(query_begin, query_begin + query_length());
+        // Position of query in concatenated text.
+        size_type const location = offset() - index->index[node.lb];
+
+        // The rank represents the number of delimiters before position `location + 1`.
+        // Furthermore, this number represents the i-th text this (text) position originates from, since each delimiter
+        // corresponds to one text in the whole concatenated text from left-to-right.
+        size_type const rank = index->text_begin_rs.rank(location + 1);
+        assert(rank > 0);
+        size_type const text_id = rank - 1;
+
+        // The sum of all text lengths prior to the i-th text is the start location of the i-th text in the concatenated
+        // sequence.
+        size_type const start_location = index->text_begin_ss.select(rank);
+        assert(text_id + 1 != 0);
```

Comment:
Does this assert make sense if you are already testing `rank > 0`?

Reply:
yeah in can be omitted